File: dsa-practice-ram/dsa-practice-ram/Arrays/MatrixZeros.cs
namespace dsa_practice_ram.Arrays
{
    internal class MatrixZeros
    {
        public void SetMatrixZerosBetterSolution()
        {
            // Mark entire row and column as zeros if there is any zero
            /*
             * 0 -> No mark, 1 -> mark 
             * 
             *   0 1 1 0
             * ---------
             0 | 1 1 1 1
             1 | 1 0 1 1
             1 | 1 1 0 1
             1 | 1 0 0 1

            Output:
                1 0 0 1
                0 0 0 0
                0 0 0 0
                0 0 0 0
             */

            int[,] matrix = new int[,]
            {
                { 1, 1, 1, 1 },
                { 1, 0, 1, 1 },
                { 1, 1, 0, 1 },
                { 1, 0, 0, 1 }
            };

            int rows = matrix.GetLength(0);
            int cols = matrix.GetLength(1);

            int[] markedRows = new int[rows];
            int[] markedCols = new int[cols];

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    if (matrix[row, col] == 0)
                    {
                        markedRows[row] = 1;
                        markedCols[col] = 1;
                    }
                }
            }

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    if (markedRows[row] == 1 || markedCols[col] == 1)
                    {
                        matrix[row, col] = 0;
                    }
                }
            }

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    Console.Write(matrix[row, col] + " ");
                }
                Console.WriteLine();
            }
        }
        
        public void SetMatrixZerosOptimal()
        {
            /* Mark entire row and column as zeros if there is any zero
             * 
             *   0|0 0 1  => It already serves as its own row marker, too.
             *  ---------
             1 | 1 1 1 1
             0 | 1 0 1 1
             0 | 1 1 0 1
             0 | 0 1 1 1

            After partially filled:

                 0|0 0 1
               ---------
             1 | * * * *
             0 | * 0 0 0
             0 | * 0 0 0
             0 | * 0 0 0

            Output:

                0 0 0 1
                0 0 0 0
                0 0 0 0
                0 0 0 0
             */

            int[,] matrix = new int[,]
            {
                { 1, 1, 1, 1 },
                { 1, 0, 1, 1 },
                { 1, 1, 0, 1 },
                { 0, 1, 1, 1 }
            };

            int rows = matrix.GetLength(0);
            int cols = matrix.GetLength(1);
            int Col0 = 1;

            // Use the first row and first column to mark
            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    if (matrix[row, col] == 0)
                    {
                        matrix[row, 0] = 0;

                        if (col == 0)
                            Col0 = 0;
                        else
                            matrix[0, col] = 0;
                    }
                }
            }

            // Update the matrix based on the markings except the first row and first column
            for (int row = 1; row < rows; row++)
            {
                for (int col = 1; col < cols; col++)
                {
                    if (matrix[row, col] != 0)
                    {
                        if (matrix[row, 0] == 0 || matrix[0, col] == 0)
                        {
                            matrix[row, col] = 0;
                        }
                    }
                }
            }

            // If the row is marked as 1 there is no need to change anything on the marked row
            if (matrix[0, 0] == 0)
            {
                for (int col = 0; col < cols; col++)
                {
                    matrix[0, col] = 0;
                }
            }

            // Update the first column based on the first element(actually saved) of the first column
            if (Col0 == 0)
            {
                for (int row = 0; row < rows; row++)
                {
                    matrix[row, 0] = 0;
                }
            }

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    Console.Write(matrix[row, col] + " ");
                }
                Console.WriteLine();
            }
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/matrix_zeros.py
"""Set matrix zeros utilities."""

# C# MatrixZeros.SetMatrixZerosBetterSolution

def set_matrix_zeros_better_solution(matrix=None):
    """Set rows and columns to zero if any element is zero (marker arrays).

    Problem: For any zero in the matrix, set its entire row and column to zero.
    Approach: First pass marks rows/cols; second pass applies zeros.
    Time complexity: O(r * c)
    Space complexity: O(r + c)
    """
    if matrix is None:
        matrix = [
            [1, 1, 1, 1],
            [1, 0, 1, 1],
            [1, 1, 0, 1],
            [1, 0, 0, 1],
        ]

    rows = len(matrix)
    cols = len(matrix[0]) if rows else 0
    marked_rows = [0] * rows
    marked_cols = [0] * cols

    for row in range(rows):
        for col in range(cols):
            if matrix[row][col] == 0:
                marked_rows[row] = 1
                marked_cols[col] = 1

    for row in range(rows):
        for col in range(cols):
            if marked_rows[row] == 1 or marked_cols[col] == 1:
                matrix[row][col] = 0

    return matrix


# C# MatrixZeros.SetMatrixZerosOptimal

def set_matrix_zeros_optimal(matrix=None):
    """Set rows and columns to zero using first row/column as markers.

    Problem: For any zero in the matrix, set its entire row and column to zero.
    Approach: Use the first row/column as markers to achieve O(1) extra space.
    Time complexity: O(r * c)
    Space complexity: O(1) extra
    """
    if matrix is None:
        matrix = [
            [1, 1, 1, 1],
            [1, 0, 1, 1],
            [1, 1, 0, 1],
            [0, 1, 1, 1],
        ]

    rows = len(matrix)
    cols = len(matrix[0]) if rows else 0
    col0 = 1

    for row in range(rows):
        for col in range(cols):
            if matrix[row][col] == 0:
                matrix[row][0] = 0
                if col == 0:
                    col0 = 0
                else:
                    matrix[0][col] = 0

    for row in range(1, rows):
        for col in range(1, cols):
            if matrix[row][col] != 0:
                if matrix[row][0] == 0 or matrix[0][col] == 0:
                    matrix[row][col] = 0

    if rows and matrix[0][0] == 0:
        for col in range(cols):
            matrix[0][col] = 0

    if col0 == 0:
        for row in range(rows):
            matrix[row][0] = 0

    return matrix
